src/client.ts

```diff
@@ -312,6 +316,125 @@ export class StreamChat<
+   * @example <caption>initialize the client in user mode</caption>
+   * StreamChat.getInstance('api_key')
+   * @example <caption>initialize the client in user mode with options</caption>
+   * StreamChat.getInstance('api_key', { timeout:5000 })
+   * @example <caption>secret is optional and only used in server side mode</caption>
+   * StreamChat.getInstance('api_key', "secret", { httpsAgent: customAgent })
+   */
+  public static getInstance<
+    AttachmentType extends UnknownType = UnknownType,
+    ChannelType extends UnknownType = UnknownType,
+    CommandType extends string = LiteralStringForUnion,
+    EventType extends UnknownType = UnknownType,
+    MessageType extends UnknownType = UnknownType,
+    ReactionType extends UnknownType = UnknownType,
+    UserType extends UnknownType = UnknownType
+  >(
+    key: string,
+    options?: StreamChatOptions,
+  ): StreamChat<
+    AttachmentType,
+    ChannelType,
+    CommandType,
+    EventType,
+    MessageType,
+    ReactionType,
+    UserType
+  >;
+  public static getInstance<
+    AttachmentType extends UnknownType = UnknownType,
+    ChannelType extends UnknownType = UnknownType,
+    CommandType extends string = LiteralStringForUnion,
+    EventType extends UnknownType = UnknownType,
+    MessageType extends UnknownType = UnknownType,
+    ReactionType extends UnknownType = UnknownType,
+    UserType extends UnknownType = UnknownType
+  >(
+    key: string,
+    secret?: string,
+    options?: StreamChatOptions,
+  ): StreamChat<
+    AttachmentType,
+    ChannelType,
+    CommandType,
+    EventType,
+    MessageType,
+    ReactionType,
+    UserType
+  >;
+  public static getInstance<
+    AttachmentType extends UnknownType = UnknownType,
+    ChannelType extends UnknownType = UnknownType,
+    CommandType extends string = LiteralStringForUnion,
+    EventType extends UnknownType = UnknownType,
+    MessageType extends UnknownType = UnknownType,
+    ReactionType extends UnknownType = UnknownType,
+    UserType extends UnknownType = UnknownType
+  >(
+    key: string,
+    secretOrOptions?: StreamChatOptions | string,
+    options?: StreamChatOptions,
+  ): StreamChat<
+    AttachmentType,
+    ChannelType,
+    CommandType,
+    EventType,
+    MessageType,
+    ReactionType,
+    UserType
+  > {
+    if (!StreamChat._instance) {
+      if (typeof secretOrOptions === 'string') {
+        StreamChat._instance = new StreamChat<
+          AttachmentType,
+          ChannelType,
+          CommandType,
+          EventType,
+          MessageType,
+          ReactionType,
+          UserType
+        >(key, secretOrOptions, options);
+      } else {
+        StreamChat._instance = new StreamChat<
+          AttachmentType,
+          ChannelType,
+          CommandType,
+          EventType,
+          MessageType,
+          ReactionType,
+          UserType
+        >(key, secretOrOptions);
+      }
```

Comment:
Do we need this if-else? StreamChat constructor handles this logic as well

Reply:
from JS perspective we don't need this, but due to fn overrides this seems the way to have a happy TS compiler